airbyte-decorator.py

```diff
@@ -0,0 +1,70 @@
+from datetime import datetime
+from typing import Any, Callable, Mapping, Optional
+
+from dagster import (
+    Definitions,
+    DefinitionsLoadContext,
+    DefinitionSource,
+    DefinitionsReloadRequest,
+    JsonMetadataValue,
+    SensorResult,
+    build_last_update_freshness_checks,
+    defs_loader,
+    sensor,
+)
+
+
+class AirbyteWorkspace:
+    def __init__(self, name: str):
+        self._name = name
+
+    def build_defs(
+        self, context: DefinitionsLoadContext, execution_fn: Optional[Callable] = None
+    ) -> Definitions:
+        existing_source: DefinitionSource = context.get_loaded_definition_source(self._name)
+        if existing_source:
+            # inside step worker
+            airbyte_workspace_info = existing_source.metadata["dagster_airbyte/workspace"].value
+        else:
+            # inside code server
+            airbyte_workspace_info = self._fetch_workspace_info()
+
+        return self._build_defs_from_workspace_info(airbyte_workspace_info)
+
+    def _build_defs_from_workspace_info(
+        self, workspace_info: Mapping[str, Any], execution_fn: Optional[Callable]
+    ) -> Definitions:
+        assets = []
+        return Definitions(
+            assets=assets,
+            sources=[
+                DefinitionSource(
+                    name=self._name,
+                    metadata={"dagster_airbyte/workspace": JsonMetadataValue(workspace_info)},
+                )
+            ],
+        )
+
+    def _fetch_workspace_info(self):
+        return {}
+
+
+workspace = AirbyteWorkspace(name="workspace1")
+
+
+@defs_loader
+def defs(context: DefinitionsLoadContext) -> Definitions:
+    airbyte_defs = workspace.build_defs(context)
```

Comment:
Note this pattern vs. the `@dbt_assets` pattern we've used in the dbt integration. The idea is that `build_defs` accepts an `execution_fn: Optional[Callable]` param that can be used if people want to customize their materialization function.  That avoids the monstrosity below, and would also generally with the problem that getting started with these integrations is intimidating.

```python
@defs_loader
def defs(context: DefinitionsLoadContext) -> Definitions:
    @workspace.airbyte_defs(context)
    def my_airbyte_defs(context)):
        yield from workspace.sync(context).stream()
    ...

```

Reply:
Yeah this is why I think for this we should consider a class-based API

Reply:
This later includes some more class-based options: https://www.notion.so/dagster/Understanding-Dagster-s-asset-integrations-ced8aa23fa93486fbc58b5ff94cf4d00?showMoveTo=true&saveParent=true